README.md
# [<img src="https://user-images.githubusercontent.com/31461850/53454621-a1b39500-39dc-11e9-9b3c-276451d42437.png" width="155px" alt="WayScript" align="center">](https://wayscript.com) JS

[![npm version](https://img.shields.io/npm/v/wayscript.svg?color=blue)](https://www.npmjs.com/package/wayscript-legacy/) [![CircleCI Status](https://circleci.com/gh/wayscript/wayscript-js-legacy/tree/master.svg?style=shield)](https://circleci.com/gh/wayscript/wayscript-js-legacy/tree/master)

### A rapid scripting platform for developers.

WayScript allows you to run Python in the cloud, and seamlessly integrate with your favorite APIs.

![Trigger scripts on any event or schedule.](https://user-images.githubusercontent.com/31461850/68791693-af5a8a80-05fe-11ea-86dd-32ccc9641bbe.png)

## Quick Start

If you use npm:

```sh
npm install wayscript-legacy
```

Or load directly from CDN:

```html
<script src="https://cdn.wayscript.com/static/js/api/wayscript-legacy.0.1.2.js"></script>
```

## Basic Usage

1. Add one or more [HTTP Triggers](https://docs.wayscript.com/library/triggers/http-trigger) to your script.

2. If you have a [password-protected endpoint](https://docs.wayscript.com/library/triggers/http-trigger#password-protect-your-endpoints), obtain your API key or the credentials you would like to use.

3. If you have specified a [custom endpoint](https://docs.wayscript.com/library/triggers/http-trigger#endpoints), you will need to pass the name of that endpoint in your api call.

4. If your HTTP Trigger takes [query parameters](https://docs.wayscript.com/library/triggers/http-trigger#request-query-parameters) and/or [JSON body parameters](https://docs.wayscript.com/library/triggers/http-trigger#request-json-body-parameters), you can pass those as a dictionary using the `params` and/or `data` arguments, respectively. (See [HTTP Trigger Outputs](https://docs.wayscript.com/library/triggers/http-trigger#outputs) for more information.)

5. Run your WayScript programs from your JavaScript code:

```javascript
wayscript = require('wayscript-legacy')

// If your program requires a password to run, supply those credentials when creating the client
wayscript.username = 'Username';
wayscript.password = 'Pa$$word';

// If your program requires a password to run, you can instead supply your API Key when creating the client
wayscript.apiKey = 'YOUR_API_KEY';

// Run a program by id
let programId = 1234;
wayscript.run( programId );

// Pass query parameters to a program (optional)
let query_params = { 'var1': 'one', 'var2': 'two', 'var3': 'three' };
wayscript.run( programId, params = query_params );

// Run a specific endpoint within your script (optional)
let endpoint = 'MyEndpoint';
wayscript.run( programId, endpoint, params = query_params );

// Pass data within the body of your request (optional)
let body_params = { 'var4': 'four', 'var5': 'five', 'bar6': 'six' };
wayscript.run( programId, endpoint, params = query_params, data = body_params );

// Handle the response
let onSuccess = function( responseText ) {
  console.log( responseText );
};
let onError = function( responseText ) {
  console.log( responseText );
};
wayscript.run( programId ).onSuccess( onSuccess ).onError( onError );
```

## Example Apps

- Node.js

    - [Run a Python script from your node app](https://github.com/wayscript/node-to-python-example)

    - [Send an email using Gmail from your node app](https://github.com/wayscript/node-to-gmail-example)
    
- JavaScript

    - [Send a Gmail message using JavaScript+WayScript](https://github.com/wayscript/js-to-gmail-example)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.1.2] - 2021-7-14
### Changed
- No-op. Now publishing to `wayscript-legacy` npm property.

## [0.1.1] - 2020-03-11
### Changed
- Move endpoint from wayscript.com to wayscript.io

## [0.1.0] - 2019-11-13
### Changed
- **BREAKING**: Convert SDK to adhere to the [WayScript HTTP Trigger](https://docs.wayscript.com/library/triggers/http-trigger) specification.

### Added
- Changelog

## [0.0.8] - 2019-07-02
### Added
- Ability to run a specific function within a program
- Link to node-to-python example in README and updated examples

## [0.0.7] - 2019-05-17
### Fixed
- Issue with passing multiple variables in an array

## [0.0.5] - 2019-04-19
### Fixed
- Example in README

## [0.0.4] - 2019-04-18
### Fixed
- URL in package.json

## [0.0.3] - 2019-04-17
### Released SDK

[Unreleased]: https://github.com/wayscript/wayscript-js-legacy/compare/0.1.2...HEAD
[0.1.2]: https://github.com/wayscript/wayscript-js-legacy/compare/0.1.1...0.1.2
[0.1.1]: https://github.com/wayscript/wayscript-js-legacy/compare/0.1.0...0.1.1
[0.1.0]: https://github.com/wayscript/wayscript-js-legacy/compare/0.0.7...0.1.0
[0.0.8]: https://github.com/wayscript/wayscript-js-legacy/compare/0.0.7...0.0.8
[0.0.7]: https://github.com/wayscript/wayscript-js-legacy/compare/0.0.5...0.0.7
[0.0.5]: https://github.com/wayscript/wayscript-js-legacy/compare/0.0.4...0.0.5
[0.0.4]: https://github.com/wayscript/wayscript-js-legacy/compare/0.0.3...0.0.4
[0.0.3]: https://github.com/wayscript/wayscript-js-legacy/releases/tag/0.0.3

test/test.js
const expect = require( 'chai' ).expect,
    wayscript = require( '../' );

describe( 'wayscript', function () {
    const DUMMY_API_KEY = "_DUMMY_API_KEY_DUMMY_API_KEY_DUMMY_API_KEY_";
    const PROGRAM_ID    = 1234;
    const QUERY_PARAMS  = { 'var1': 'one,', 'var2': 'two&', 'var3': 'three?' };
    const BODY_PARAMS   = { 'var4': 'four,', 'var5': 'five?', 'var6': 'six:' };

    it( 'should allow api key assignment', function () {
        expect( wayscript.apiKey ).to.be.equal( '' );
        wayscript.apiKey = 'foobar';
        expect( wayscript.apiKey ).to.be.equal( 'foobar' );
    } );

    it( 'should allow username and password assignment', function () {
        expect( wayscript.username ).to.be.equal( '' );
        expect( wayscript.password ).to.be.equal( '' );
        wayscript.username = 'captain@wayscript.com';
        wayscript.password = 'letmein';
        expect( wayscript.username ).to.be.equal( 'captain@wayscript.com' );
        expect( wayscript.password ).to.be.equal( 'letmein' );
    } );

    it( 'should throw on invalid api key', function () {
        wayscript.apiKey = 'foobar';
        expect( wayscript.apiKey ).to.be.equal( 'foobar' );
        expect( function () {
            wayscript.run();
        } ).to.throw( 'The API Key provided is not valid.' );
    } );

    it( 'should run program', function () {
        wayscript.apiKey = DUMMY_API_KEY;
        expect( function () {
            wayscript.run( PROGRAM_ID );
        } ).to.not.throw();
    } );

    it( 'should run program with query params', function () {
        expect( function () {
            wayscript.run( PROGRAM_ID, params = QUERY_PARAMS );
        } ).to.not.throw();
    } );

    it( 'should run program with empty query params', function () {
        expect( function () {
            wayscript.run( PROGRAM_ID, {} );
        } ).to.not.throw();
    } );

    it( 'should run program with body params', function () {
        expect( function () {
            wayscript.run( PROGRAM_ID, data = BODY_PARAMS );
        } ).to.not.throw();
    } );

    it( 'should run program with empty body params', function () {
        expect( function () {
            wayscript.run( PROGRAM_ID, {} );
        } ).to.not.throw();
    } );

    it( 'should run program with endpoint', function () {
        expect( function () {
            wayscript.run( PROGRAM_ID, 'My%20Function' );
        } ).to.not.throw();
    } );

    it( 'should run program with endpoint and params', function () {
        expect( function () {
            wayscript.run( PROGRAM_ID, 'My%20Function', params = QUERY_PARAMS, data = BODY_PARAMS );
        } ).to.not.throw();
    } );

    it( 'should allow onerror/onsuccess assignment', function ( done ) {
        const onError = function onError( text ) {
            expect( text ).to.contain( '{"Error":"You must copy this script in order for it to run."}' );
            done();
        };

        const onSuccess = function onSuccess( text ) {
        };

        let response = wayscript.run( PROGRAM_ID, 'My Function', params = QUERY_PARAMS, data = BODY_PARAMS )
            .onError( onError )
            .onSuccess( onSuccess );

        expect( response._onError ).to.be.equal( onError );
        expect( response._onSuccess ).to.be.equal( onSuccess );
        expect( response.authorizationHeader ).to.be.equal( 'Bearer _DUMMY_API_KEY_DUMMY_API_KEY_DUMMY_API_KEY_' );
        expect( response.requestUrl ).to.be.equal( 'https://1234.wayscript.io/My%20Function?var1=one%2C&var2=two%26&var3=three%3F' )
        expect( response.requestParams ).to.be.equal( '?var1=one%2C&var2=two%26&var3=three%3F' );
        expect( response.requestBody ).to.be.equal( BODY_PARAMS );
    } );
} );


js/wayscript.js
/**
 * --------------------------------------------------------------------------
 * WayScript, Inc.: wayscript.js
 * Licensed under MIT (https://github.com/wayscript/wayscript-js/blob/master/LICENSE)
 * --------------------------------------------------------------------------
 */

( function ( root, factory ) {
    if ( typeof define === 'function' && define.amd ) {
        define( factory( root ) );
    } else if ( typeof exports === 'object' ) {
        module.exports = factory( require( 'xmlhttprequest' ) );
    } else {
        root.wayscript = factory( root );
    }
} )( typeof global !== "undefined" ? global : this.window || this.global, function ( root ) {
    'use strict';

    const wayScript = {};

    wayScript.apiKey = '';

    wayScript.runProgram = function ( programId, variables ) {
        if ( !wayScript.apiKey || wayScript.apiKey.length !== 43 ){
            throw new Error( 'The API Key provided is not valid.' );
        }

        let params = '?api_key=' + wayScript.apiKey + '&program_id=' + programId;

        if ( variables && variables.length ) {
            params += '&variables=' + variables;
        }

        return _post( params );
    };

    const _post = function ( params ) {
        let xhr = new root.XMLHttpRequest();
        xhr.open( "POST", 'https://wayscript.com/api' + params );
        xhr.setRequestHeader( "Content-Type", "application/json" );
        xhr.setRequestHeader( "X-WayScript-Api", "javascript" );

        const response = {};

        response.onSuccess = function( func ) {
            response._onSuccess = func;
            return response;
        };

        response.onError = function( func ) {
            response._onError = func;
            return response;
        };

        xhr.onreadystatechange = function () {
            if ( xhr.readyState === 4 ) {
                if ( xhr.status === 200 ) {
                    if ( response._onSuccess ) {
                        response._onSuccess( xhr.responseText );
                    }
                } else if ( response._onError ) {
                    response._onError( xhr.responseText );
                }
            }
        };

        xhr.send();

        return response;
    };

    return wayScript;
} );


index.js
/**
 * --------------------------------------------------------------------------
 * WayScript, Inc.: wayscript.js
 * Licensed under MIT (https://github.com/wayscript/wayscript-js/blob/master/LICENSE)
 * --------------------------------------------------------------------------
 */

( function ( root, factory ) {
    if ( typeof define === 'function' && define.amd ) {
        define( [ 'xmlhttprequest', 'buffer' ], factory  );
    } else if ( typeof exports === 'object' ) {
        module.exports = factory( require( 'xmlhttprequest' ).XMLHttpRequest, require( 'buffer' ).Buffer );
    } else {
        root.wayscript = factory( root.XMLHttpRequest, root.Buffer );
    }
} )( typeof global !== "undefined" ? global : this.window || this.global, function ( XMLHttpRequest, Buffer ) {
    'use strict';

    const wayScript = {};

    wayScript.apiKey = '';
    wayScript.username = '';
    wayScript.password = '';

    wayScript.run = function ( programId, endpoint = '', params = null, data = null ) {
        let query_param_str = '';
        endpoint = encodeURIComponent( endpoint || '' );

        if ( params ) {
            for ( let [key, value] of Object.entries( params ) ) {
                query_param_str += query_param_str.length > 0 ? '&' : '?';
                query_param_str += encodeURIComponent( key ) + '=' + encodeURIComponent( value );
            }
        }

        return _post( programId, endpoint, query_param_str, data );
    };

    const _post = function ( program_id, endpoint, query_param_str, body_params ) {
        let xhr = new XMLHttpRequest();
        const request_url = 'https://' + program_id + '.wayscript.io/' + endpoint + query_param_str;
        xhr.open( "POST", request_url );
        xhr.setRequestHeader( "Content-Type", "application/json" );
        xhr.setRequestHeader( "X-WayScript-Api", "javascript" );

        const auth_header = _get_auth_header();
        if ( auth_header ) xhr.setRequestHeader( "Authorization", auth_header );

        const response = {};
        response.requestUrl = request_url;
        response.authorizationHeader = auth_header;
        response.requestParams = query_param_str;
        response.requestBody = body_params;

        response.onSuccess = function( func ) {
            response._onSuccess = func;
            return response;
        };

        response.onError = function( func ) {
            response._onError = func;
            return response;
        };

        xhr.onreadystatechange = function () {
            if ( xhr.readyState === 4 ) {
                if ( xhr.status === 200 ) {
                    if ( response._onSuccess ) {
                        response._onSuccess( xhr.responseText );
                    }
                } else if ( response._onError ) {
                    response._onError( xhr.responseText );
                }
            }
        };

        xhr.send( JSON.stringify( body_params || { } ) );

        return response;
    };

    const _get_auth_header = function() {
        if ( wayScript.apiKey ) {
            if ( wayScript.apiKey.length !== 43 ) {
                throw new Error( 'The API Key provided is not valid.' );
            }
            return 'Bearer ' + wayScript.apiKey;
        }
        else if ( wayScript.username && wayScript.password ) {
            return 'Basic ' + _get_base64_string( wayScript.username + ':' + wayScript.password );
        }
        return null;
    };

    const _get_base64_string = function( str ) {
        if ( typeof btoa === "function" ) {
            return btoa( str );
        } else {
            return Buffer.from( str ).toString( 'base64' );
        }
    };

    return wayScript;
} );
